AlexDegterov/shop-be
# Task 9

[Task 9](https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/task9-bff-elastic-beanstalk/task.md)

**link to the repo (back-end)**
https://github.com/AlexDegterov/shop-be

**link to the repo (front-end)**
https://github.com/AlexDegterov/shop-vue-vuex-cloudfront

**product-service service API endpoint URL**
Все товары - https://4eumx86rj2.execute-api.eu-west-1.amazonaws.com/dev/products
Информация по конкретному товару - https://4eumx86rj2.execute-api.eu-west-1.amazonaws.com/dev/products/429b4c59-efd0-4c94-8097-a9b72675cc89

**example of the create product API call with all needed information: URL, payload, headers, etc.**
На странице https://d35wgkj3ohjo44.cloudfront.net/admin/products можно загрузить csv-файл следующего формата:
> title,description,price,logo,count
"Тестовый продукт для проверки","Тестовое описание для проверки",100,"https://www.ekk.by/files/files/page/366/pv-3-sh.jpg",500

При этом необходимо создать в Local Storage ключ `authorization_token` со значением `'Basic 1QWxleERlZ3Rlcm92OlRFU1RfUEFTU1dPUkQ='` для авторизации.
После загрузки товара на странице https://4eumx86rj2.execute-api.eu-west-1.amazonaws.com/dev/products будет выдаваться информация сразу, а на странице (bff) http://alexdegterov-bff-api-dev.eu-west-1.elasticbeanstalk.com/product выдаваться закэшированная информация (2 минуты). Если данные будут совпадать, то необходимо:
- сначала зайти на страницу  http://alexdegterov-bff-api-dev.eu-west-1.elasticbeanstalk.com/product
- потом добавить csv-файл
- потом проверять снова (bff) http://alexdegterov-bff-api-dev.eu-west-1.elasticbeanstalk.com/product (выдаст старый кэш без нового товара) и страницу https://4eumx86rj2.execute-api.eu-west-1.amazonaws.com/dev/products (выдаст с новым товаром, без кэширования)

**CART service API endpoint URL**
Корзина товаров на прямую - https://lg11pm0xhk.execute-api.eu-west-1.amazonaws.com/dev/api/profile/cart

**bff-service service URL**
Общий адрес (выдает ошибку 502 как по заданию, т.к. нет запроса к конкретному сервису) - http://alexdegterov-bff-api-dev.eu-west-1.elasticbeanstalk.com/

**example how to call product-service and CART services via bff-service service URL**
Все товары через bff (кэш 2 минуты) - http://alexdegterov-bff-api-dev.eu-west-1.elasticbeanstalk.com/product
Информация по конкретному товару через bff - http://alexdegterov-bff-api-dev.eu-west-1.elasticbeanstalk.com/product/?id=429b4c59-efd0-4c94-8097-a9b72675cc89
Корзина товаров через bff (вне задания, кэш 2 минуты) - http://alexdegterov-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart

### Evaluation criteria:
   **ГОТОВО** 3 - A working and correct express application should be in the bff-service folder. Reviewer can start this application locally with any valid configuration in the .env file and this application should works as described in the task 9.1
   **ГОТОВО** 5 - The bff-service should be deployed with Elastic Beanstalk. The bff-service call should be redirected to the appropriate service : product-service or CART. The response from the bff-service should be the same as if product-service or CART services were called directly.

### Additional (optional) tasks
   **ГОТОВО** +1 - Add a cache at the bff-service level for a request to the getProductsList function of the product-service. The cache should expire in 2 minutes.
    How to test:
        Get products list
        Create new product
        Get products list - the result shouldn’t have a new product
        Wait more than 2 minutes
        Get products list - the result should have a new product

### Итого: 5 + 1 = 6 баллов
_Спасибо за Вашу оценку. Если не согласны с оценкой в 6 баллов, то прошу написать мне в Discrord - AlexD#4666._
